synapse/handlers/message.py

```diff
@@ -968,6 +971,13 @@ async def create_new_client_event(
         if builder.internal_metadata.outlier:
             event.internal_metadata.outlier = True
             context = EventContext.for_outlier()
+        elif (
+            event.type == EventTypes.MSC2716_INSERTION
+            and full_state_ids_at_event
+            and builder.internal_metadata.is_historical()
+        ):
+            old_state = await self.store.get_events_as_list(full_state_ids_at_event)
+            context = await self.state.compute_event_context(event, old_state=old_state)
```

Comment:
Add explicit state to the insertion event so the rest of the batch can inherit the same state and `state_group`